Bluetooth: GATT: Fix not clearing stored data when unpairing

Vudentz opened the pull request:
GATT data shall not be considered conditional to BT_SETTINGS since
the data is stored in RAM it must also be cleared when unpairing.

Fixes #22514

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Move functions related to bt_gatt_clear

These functions shall not be conditional to BT_SETTINGS.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
@ivan98ams Give this a try, it should fix #22514

ivan98ams commented:
Works for me!

jhedberg commented:
@Vudentz I think there are several places where you can get rid of #ifdefs from around or inside static functions, by making sure the calls to those functions are protected by IS_ENABLED()

joerchan commented:
@Vudentz Is it possible to split the commit into 2 commits? One for moving the functions inside the file, and one for the actual bug-fix. The diff is a bit hard to read at the moment.

Vudentz committed:
Bluetooth: GATT: Fix not clearing stored data when unpairing

GATT data shall not be considered conditional to BT_SETTINGS since
the data is stored in RAM it must also be cleared when unpairing.

Fixes #22514

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
> @Vudentz I think there are several places where you can get rid of #ifdefs from around or inside static functions, by making sure the calls to those functions are protected by IS_ENABLED()

This is now done, there were a few dependencies to variable declared only inside a #if defined so I needed to a little bit more than just replace #if by IS_ENABLED.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -4473,3 +4348,154 @@ SETTINGS_STATIC_HANDLER_DEFINE(bt_hash, "bt/hash", NULL, db_hash_set,
+		return err;
+	}
+
+	bt_gatt_clear_subscriptions(addr);
```
Why not put `if (IS_ENABLED(CONFIG_BT_GATT_CLIENT))` here instead of doing it inside the function definition. That way you don't need to indent so much code.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -4473,3 +4348,154 @@ SETTINGS_STATIC_HANDLER_DEFINE(bt_hash, "bt/hash", NULL, db_hash_set,
+	err = sc_clear_by_addr(id, addr);
+	if (err < 0) {
+		return err;
+	}
```
You have pretty long `if (IS_ENABLED())` branches inside `bt_gatt_clear_ccc()` and `sc_clear_by_addr()`. Why not put the IS_ENABLED() here instead so you don't need to indent so much code?

Vudentz replied:
Right, I had it replacing the #if location but that indeed messes with identation a little too much.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -4473,3 +4348,154 @@ SETTINGS_STATIC_HANDLER_DEFINE(bt_hash, "bt/hash", NULL, db_hash_set,
+	err = bt_gatt_clear_cf(id, addr);
+	if (err < 0) {
+		return err;
+	}
```
Same here

jhedberg reviewed (changes requested):
@Vudentz looks better, but I think you can move the IS_ENABLED() checks one call higher up to avoid indenting so much code.

Vudentz commented:
> @Vudentz looks better, but I think you can move the IS_ENABLED() checks one call higher up to avoid indenting so much code.

Done.

jhedberg commented:
@Vudentz please still fix the checkpatch warning regarding redundant return statement

Vudentz commented:
> @Vudentz please still fix the checkpatch warning regarding redundant return statement

Done.

Vudentz committed:
Bluetooth: GATT: Fix not clearing SC when disconnecting

SC config data is no longer stored within the CCC config itself
therefore it must be cleared separately.

Fixes #22539

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
@chris-makaio Ive added a for dealing with #22539 as it is somewhat related though you were not using bt_unpair but a regular disconnect which still was not clearing thing properly.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -4473,3 +4355,154 @@ SETTINGS_STATIC_HANDLER_DEFINE(bt_hash, "bt/hash", NULL, db_hash_set,
+
+			u8_to_dec(id_str, sizeof(id_str), id);
+			bt_settings_encode_key(key, sizeof(key), "ccc",
+					       (bt_addr_le_t *)addr, id_str);
```
Unrelated to this PR, but it looks like `bt_settings_encode_key()` needs to be fixed to take a const address pointer

jhedberg replied:
Actually, I take that back. @Vudentz could you include the following patch as a new commit into this PR, and then fix up your current commit not to do the explicit casts:
```
diff --git a/subsys/bluetooth/host/settings.c b/subsys/bluetooth/host/settings.c
index a97bfe9bec..ea138efaf4 100644
--- a/subsys/bluetooth/host/settings.c
+++ b/subsys/bluetooth/host/settings.c
@@ -21,7 +21,7 @@
 
 #if defined(CONFIG_BT_SETTINGS_USE_PRINTK)
 void bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
-                           bt_addr_le_t *addr, const char *key)
+                           const bt_addr_le_t *addr, const char *key)
 {
        if (key) {
                snprintk(path, path_size,
@@ -41,7 +41,7 @@ void bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
 }
 #else
 void bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
-                           bt_addr_le_t *addr, const char *key)
+                           const bt_addr_le_t *addr, const char *key)
 {
        size_t len = 3;
 
diff --git a/subsys/bluetooth/host/settings.h b/subsys/bluetooth/host/settings.h
index c918ad426c..e4854e437c 100644
--- a/subsys/bluetooth/host/settings.h
+++ b/subsys/bluetooth/host/settings.h
@@ -12,7 +12,7 @@
 
 /* Helpers for keys containing a bdaddr */
 void bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
-                           bt_addr_le_t *addr, const char *key);
+                           const bt_addr_le_t *addr, const char *key);
 int bt_settings_decode_key(const char *key, bt_addr_le_t *addr);
 
 void bt_settings_save_id(void);
```

joerchan commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2131,6 +2136,9 @@ static u8_t disconnected_cb(const struct bt_gatt_attr *attr, void *user_data)
 			if (!bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
+				if (ccc == &sc_ccc) {
+					sc_clear_by_conn(conn);
+				}
```
style: blank line.

joerchan commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2131,6 +2136,9 @@ static u8_t disconnected_cb(const struct bt_gatt_attr *attr, void *user_data)
 		} else {
 			/* Clear value if not paired */
 			if (!bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
+				if (ccc == &sc_ccc) {
```
Good find!

joerchan commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3693,7 +3693,6 @@ static void gatt_sub_remove(struct bt_conn *conn, struct gatt_sub *sub,
 			    sys_snode_t *prev,
 			    struct bt_gatt_subscribe_params *params)
 {
-	return;
```
This could be squashed to the previous commit.

joerchan reviewed (commented):
Looks good, got a few nitpicks.

Finding the second bug was good. I had a note about SC changed issue that i couldn't figure out what I meant anymore. So happy that you found it.

Vudentz committed:
Bluetooth: Setting: Make bt_settings_encode_key take a const address

This removes the need to cast in case the address is already const.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>
